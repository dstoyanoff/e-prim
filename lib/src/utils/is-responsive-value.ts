import { ResponsiveValue } from "../system/types";
import { TBreakpoint } from "../theme";

/**
 * Check if the value is an object that contains either breakpoint
 */
export const isResponsiveValue = <T>(value: ResponsiveValue<T>, breakpoints: TBreakpoint) => {
  if (typeof value !== "object") {
    return false;
  }

  return Object.keys(breakpoints).some(breakpoint => value && breakpoint in value);
};
