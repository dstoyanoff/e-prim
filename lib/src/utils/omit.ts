export const omit = <T extends object>(obj: T, ...keys: (keyof T)[]) => {
  return Object.entries(obj).reduce((acc, [key, value]) => {
    if (!keys.includes(key as keyof T)) {
      acc[key] = value;
    }

    return acc;
  }, {} as Record<string, unknown>) as T;
};
