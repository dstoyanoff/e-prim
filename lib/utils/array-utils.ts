/**
 * Takes either a single element or an array and always returns an array of one or many elements
 * @param items input items of generic type
 */
export const singleOrMulti = <T>(items: T | T[]): T[] => (Array.isArray(items) ? items : [items]);

type ArrayLengthMutationKeys = "splice" | "push" | "pop" | "shift" | "unshift";

export type FixedLengthArray<T, L extends number, TObj = [T, ...Array<T>]> = Pick<
  TObj,
  Exclude<keyof TObj, ArrayLengthMutationKeys>
> & {
  readonly length: L;
  [I: number]: T;
  [Symbol.iterator]: () => IterableIterator<T>;
};
