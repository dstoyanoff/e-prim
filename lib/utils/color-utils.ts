/**
 * Converts a decimal percent to hex
 * @param p percent to convert. Should be an integer between 0 and 100, other values are bounded to that range.
 * Decimal values are rounded to the nearest integer
 */
export const percentToHex = (p: number) => {
  const percent = Math.max(0, Math.min(100, p)); // bound percent from 0 to 100
  const intValue = Math.round((percent / 100) * 255); // map percent to nearest integer (0 - 255)
  const hexValue = intValue.toString(16); // get hexadecimal representation

  return hexValue.padStart(2, "0"); // format with leading 0
};
