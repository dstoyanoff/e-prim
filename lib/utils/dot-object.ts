export type PathsToStringProps<T, TStopType = string> = T extends TStopType
  ? []
  : {
      [K in Extract<keyof T, string | number>]: [
        K,
        ...PathsToStringProps<T[K], TStopType>
      ];
    }[Extract<keyof T, string | number>];

export type Join<
  T extends (string | number)[],
  D extends string = "."
> = T extends []
  ? never
  : T extends [infer F]
  ? F
  : T extends [infer F, ...infer R]
  ? F extends string | number
    ? `${F}${D}${Join<Extract<R, (string | number)[]>, D>}`
    : never
  : string | number;

/**
 * Given a color key (dot-based deep path), resolves the actual color value
 * @param key dot-based path
 * @param object reference to the main object that should be used to find the value
 */
export const getValueFromKey = <
  TValue,
  TKey extends string = string,
  TObject = {}
>(
  key: TKey | undefined,
  object: TObject
): TValue | undefined => {
  if (!key) {
    return undefined;
  }

  const paths = key.split(".");

  return paths.reduce((accumulator, path) => {
    return accumulator ? accumulator[path] : (object as any)[path];
  }, undefined);
};
