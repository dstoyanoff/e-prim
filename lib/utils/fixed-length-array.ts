type ArrayLengthMutationKeys = "splice" | "push" | "pop" | "shift" | "unshift";

export type FixedLengthArray<T, L extends number, TObj = [T, ...Array<T>]> = Pick<
  TObj,
  Exclude<keyof TObj, ArrayLengthMutationKeys>
> & {
  readonly length: L;
  [I: number]: T;
  [Symbol.iterator]: () => IterableIterator<T>;
};
