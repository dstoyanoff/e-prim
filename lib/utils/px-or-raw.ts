export const pxOrRaw = (value: string | number, numberTransform: (value: number) => number = v => v) =>
  typeof value === "number" ? `${numberTransform(value)}px` : value;
