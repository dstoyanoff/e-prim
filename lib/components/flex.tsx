import { ElementType, forwardRef, ReactElement, Ref } from "react";
import { combineResponsiveValues } from "@/system/combine-responsive-values";
import { createFlex, flexPropKeys, FlexSystem } from "@/system/flex-system";
import { omit } from "@/utils/omit";
import { Box, BoxProps } from "./box";

const DEFAULT_TAG = "div";

type OwnProps = FlexSystem;

export type FlexProps<E extends ElementType> = BoxProps<E> &
  OwnProps & {
    as?: E;
  };

/**
 * The Flex component is a primitive that exposes useful properties for faster prototyping of flex layouts
 */
export const Flex: <E extends ElementType = typeof DEFAULT_TAG>(props: FlexProps<E>) => ReactElement | null =
  forwardRef(<E extends ElementType = typeof DEFAULT_TAG>(props: FlexProps<E>, ref: Ref<FlexProps<E>["as"]>) => (
    <Box
      ref={ref}
      {...omit(props, ...flexPropKeys)}
      css={theme => combineResponsiveValues(...createFlex(props, theme))}
    />
  ));
