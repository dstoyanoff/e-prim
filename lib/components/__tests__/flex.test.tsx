/**
 * @jest-environment jsdom
 */

import { render } from "@testing-library/react";
import { wrapper } from "@/utils/mock-theme";
import { Flex } from "../flex";

describe("flex", () => {
  test("should render flex", () => {
    expect(
      render(
        <Flex
          direction="column"
          gap={{ xs: 1, md: 4 }}
          color="neutral.4"
          background="primary.normal"
          typography="body.3"
          width={500}
          center
          height="50%"
          p={5}
          grow={1}
        >
          child
        </Flex>,
        {
          wrapper,
        },
      ).container,
    ).toMatchSnapshot();
  });
});
