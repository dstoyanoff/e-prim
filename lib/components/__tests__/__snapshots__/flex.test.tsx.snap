// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flex should render flex 1`] = `
.emotion-0 {
  width: 500px;
  height: 50%;
  color: #898896;
  background: #00659e;
  font-family: Mulish;
  font-size: 11px;
  font-weight: 400;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  padding: 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

@media (min-width: 0px) {
  .emotion-0 {
    gap: 4px;
  }
}

@media (min-width: 500px) {
  .emotion-0 {
    gap: 16px;
  }
}

<div>
  <div
    class="emotion-0"
  >
    child
  </div>
</div>
`;
